### PR TITLE
Fix #39: Using Salt Formulas 

### LINKED ISSUE #39: Add section in automation documentation page with the required steps to run the formulas
Hello,

In order to improve the documentation of the automation page, we should add a new section explaining how to use the formulas in a generic way. I have the next in my mind.

Create a new section after all of the formulas are listed. Something like `How to use the formulas`.
This section would have the next information:

The salt formulas can be used with 2 different approaches: salt master/minion or only salt minion execution.

For the master/minion approach, all of the next steps must be executed in the master machines. If only salt minion option is used, the steps must be executed in all of the minions where the formulas are going to executed.

1. Install the formulas
Install the formulas that are going to be applied using zypper

2. Create the pillar files structure
The pillar files work as configuration files for the salt formulas, they are the input the user has to give. In order to use the files, the next structure is the most appropriate one (this is only an example):

a. Create the `/srv/pillar` folder (this might be already in place)
b. Create a `top.sls` file in `/srv/pillar` with the next content. This will apply the `hana`, `netweaver`, `drbd` and `ha` formulas in the nodes specified by the hostname.

```
base:
    'hana01,hana02':
        - hana.hana
        - hana.cluster

    'netweaver01,netweaver02':
        - netweaver.netweaver
        - netweaver.cluster

    'drbd01,drbd02':
        - drbd.drbd
        - drbd.cluster
```

c. Create the folders `/srv/pillar/hana`, `/srv/pillar/netweaver` and `/srv/pillar/drbd` for the pillar files (notice that these names match with the names used in the `top.sls` file.

d. Add the pillar files in the correct folder. `hana` folder will have `hana.sls` and `cluster.sls`, `netweaver` folder will have `netweaver.sls` and `cluster.sls`, and `drbd` folder will have `drbd.sls` and `cluster.sls`.

Find here an example of the contents you can use for each of the formulas:
[hana.sls](https://github.com/SUSE/saphanabootstrap-formula/blob/master/pillar.example)
[netwaever.sls](https://github.com/SUSE/sapnwbootstrap-formula/blob/master/pillar.example)
[drbd.sls](https://github.com/SUSE/drbd-formula/blob/master/pillar.example)
[cluster.sls](https://github.com/SUSE/habootstrap-formula/blob/master/pillar.example)

The content of the pillar files must be configured depending on the needed configuration of each of the formulas (notice that in this example, 3 different `cluster.sls` are used, and each of them might be different, as the HA clusters requirements might be different too.

e. Create a `top.sls` in `/srv/salt` folder. It follows the same rules as the pillar `top.sls` file. Here an example (in this case the folder part is removed, as we are pointing to the formulas with their name):

```
base:
    'hana01,hana02':
        - hana
        - cluster

    'netweaver01,netweaver02':
        - netweaver
        - cluster

    'drbd01,drbd02':
        - drbd
        - cluster
```

3. Execute salt for each of the approaches (salt supports more options than those, this is just an example)
a. For the master/minion approach run:
```
salt '*' state.highstate
```
b. For the only minion approach run:
```
salt-call --local state.highstate
```

---

Edit toms:
Related to [bsc#1174530](https://bugzilla.suse.com/show_bug.cgi?id=1174530)

### PR DESCRIPTION
**Pull request by tomschr**
This PR contains:

* a new section "Using Salt Formulas" and a procedure on how to apply formulas
* describes both ways: Salt master/minion and Salt minion only

PDF with remarks:
[art-sol-automation_remarks_color_draft_en.pdf](https://github.com/SUSE/doc-slesforsap/files/5266711/art-sol-automation_remarks_color_draft_en.pdf)

------
@arbulu89 Would you like to have a look? :wink: I've added some remarks and questions, maybe you could answer them in the review. See page 4ff.

I took the freedom to change some steps slightly from your original post. The reason for this was to make it more consistent and to make it easier to read.

### CODE REVIEW

File: xml/art_solution_automation.xml [hunk 1]
```diff
@@ -166,6 +166,175 @@ https://confluence.suse.com/x/8IEnGw
...
+            executed in all of the &salt; clients where the formulas are going to executed.
+          </para>
+        </formalpara>
+      </listitem>
+    </itemizedlist>
+
+    <procedure xml:id="pro-sol-autom-hana-apply-salt-formulas">
+      <title>Applying &salt; Formulas</title>
+      <para>
+        In this procedure we show how an example how to create a pillar structure for
```
Comment (arbulu89) [inline xml/art_solution_automation.xml:197]: It looks like something is wrong in this paragraph

File: xml/art_solution_automation.xml [hunk 2]
```diff
@@ -166,6 +166,175 @@ https://confluence.suse.com/x/8IEnGw
...
+        a system with &hana;, &netweaver;, and with DRBD. All this uses &sleha;.
```
Comment (arbulu89) [inline xml/art_solution_automation.xml:198]: I don't know if is correct to mention the High Availability extension here. We should say `SLES4SAP` as some of the packages are coming here.

File: xml/art_solution_automation.xml [hunk 3]
```diff
@@ -166,6 +166,175 @@ https://confluence.suse.com/x/8IEnGw
...
+        </substeps>
+      </step>
+      <step>
+        <para>
+          Download the pillar files and store them in the correct directory under <filename
+            class="directory">/srv/pillar</filename>:
+        </para>
+        <substeps>
+          <step>
+            <remark>toms 2020-09-23: You can download the files and store them
```
Comment (arbulu89) [inline xml/art_solution_automation.xml:256]: I don't know what to say... One or two steps looks the same to me.
Actually, I would just say that the pillar configuration files must be created, and downloading them from the github repositories is one option. In any case.
The current text looks good to me

File: xml/art_solution_automation.xml [hunk 4]
```diff
@@ -166,6 +166,175 @@ https://confluence.suse.com/x/8IEnGw
...
+        <para>
+          Modify the content of the pillar files depending on your needed configuration
+          of each &salt; formula.
+        </para>
+        <para>
+          Notice, in this example, we have three different
+          <filename>cluster.sls</filename> files. Each might have different HA cluster
+          requirements.
+        </para>
+        <remark>toms 2020-09-23: Maybe we should link to the upstream Salt documentation here?
```
Comment (arbulu89) [inline xml/art_solution_automation.xml:293]: I don't think referencing any salt documentation will help. Here the thing is that the users must know which are the things they need to configure (and that's the difficult part).
As we commented, we cannot give already done pillar files in the documentation. Maybe we could find other source, but in any case we will always have corner case.

File: xml/art_solution_automation.xml [hunk 5]
```diff
@@ -166,6 +166,175 @@ https://confluence.suse.com/x/8IEnGw
...
+        - netweaver
+        - cluster
+
+    'drbd01,drbd02':
+        - drbd
+        - cluster</screen>
+        <para>
+          In the above example, folder part is removed, as we are pointing to the formulas with
+          their name.
+          <remark>toms 2020-09-23: why? Could we make it consistent with /srv/pillar/top.sls?
```
Comment (arbulu89) [inline xml/art_solution_automation.xml:316]: No. I think this it the best way.
As we have 3 different HA cluster, with their own `cluster.sls` file, the idea of using folders looks the best for the pillars.
The salt states are different, we only have one formula, and this is the name they must use.

File: xml/art_solution_automation.xml [hunk 6]
```diff
@@ -166,6 +166,175 @@ https://confluence.suse.com/x/8IEnGw
...
+          <step>
+            <para>If you use a &saltmaster; with &salt; client, run:</para>
+            <screen>&prompt.root;<command>salt</command> '*' state.highstate</screen>
+          </step>
+          <step>
+            <para>If you use a &salt; client only, run:</para>
+            <screen>&prompt.root;<command>salt</command> --local state.highstate</screen>
+          </step>
+        </stepalternatives>
+        <remark>toms 2020-09-23: I miss some kind of "closing" step. How can I see/check
```
Comment (arbulu89) [inline xml/art_solution_automation.xml:333]: The salt execution will report the status of the deployment. If everything is green (literally), everything went fine. If we have some state with error (in red), this means that something went wrong.

**Review by arbulu89 — COMMENTED** [review]
@tomschr I think that the new chapter in general is really well done. I have caught some potential errors and added comments.

I commented all the remarks (I think).

As always (and this will happen in the SUSE Manager SAP guide too), the most difficult part of this process is the pillar files configuration. We expect that the users know what they are doing.
This guide is not meant for `copy/paste` process in my opinion. If this is what users want, we need to change the scope, and create step by step guides, or provide the `best practices` configuration files somehow. But this is not as easy as it looks like, as the users want some customization after all

File: xml/art_solution_automation.xml (hunk 1 above)
Comment (tomschr) [reply xml/art_solution_automation.xml:197]: Argh, yes, you are right. :+1:

File: xml/art_solution_automation.xml (hunk 2 above)
Comment (tomschr) [reply xml/art_solution_automation.xml:198]: That correct. I limited that to SLEHA only, but you are right I should better use SLES4SAP here. Fixed.

File: xml/art_solution_automation.xml (hunk 4 above)
Comment (tomschr) [reply xml/art_solution_automation.xml:293]: Right. It was more of an idea. I think, I can solve that by adding an additional paragraph *before* the procedure:

```
The following procedure is written in a general way to make it fit for all
use cases. You need some knowledge about &salt; and how to
customize it to your needs.
```

File: xml/art_solution_automation.xml (hunk 3 above)
Comment (tomschr) [reply xml/art_solution_automation.xml:256]: Hmn... :thinking: ... did I get that right, you think the text "Download the pillar files and store them in the correct directory under `/srv/pillar/`" would be enough?

File: xml/art_solution_automation.xml (hunk 5 above)
Comment (tomschr) [reply xml/art_solution_automation.xml:316]: Ok, if you think that's the way how it should be, then I'm fine. :-) Removing the remark.

File: xml/art_solution_automation.xml (hunk 6 above)
Comment (tomschr) [reply xml/art_solution_automation.xml:333]: Ahh great! :+1: I've replaced the remark with:

```
The command shows you the result (green for everything is fine, red for errors).
```

File: xml/art_solution_automation.xml (hunk 3 above)
Comment (arbulu89) [reply xml/art_solution_automation.xml:256]: What I mean is that the user needs to create the pillar files and put the content. Downloading them from the current examples is one option (maybe the best option by now). I just wanted to emphisize (that they don't necessarily need to download). It think the current text is good.

Comment (tomschr) [reply xml/art_solution_automation.xml]: I've tried to rearrange these substeps into steps, I think that makes it clearer. Did that in commit fd9bd33.

Comment (tomschr) [reply xml/art_solution_automation.xml]: > Downloading them from the current examples is one option

Ahh, I see. I'm thinking about if I should mention this or better leave it as is (to not complicate things). :thinking: I think, I leave it for the time being.

**Comment by tomschr** [comment]
@arbulu89 Thank you very much for your review! Much appreciated! :+1:  I've worked on all of your comments and removed all remarks. Hope I haven't overlooked anything.

You can find the result here:
[art-sol-automation_remarks_color_draft_en.pdf](https://github.com/SUSE/doc-slesforsap/files/5268176/art-sol-automation_remarks_color_draft_en.pdf)

If you are okay with the result, let me know so I can merge it. :wink:  Thanks!

**Comment by Loquacity** [comment]
Please check https://github.com/uyuni-project/uyuni-docs/pull/559

**Comment by tomschr** [comment]
Talked to Lana today: we come to the conclusion that she will add a link in her guide pointing to this doc/section.

**Review by arbulu89 — APPROVED** [review]
@tomschr After the changes, it looks fine to me! Good work.

**Comment by MalloZup** [comment]
ok that is great news! thx @tomschr  for helping lana.
Doc is good. 

However I do think right now `zypper install saphanabootstrap-formula` is not really the documentation for "using" the formula on its scenario and combination.

What is our take to document this: 
https://github.com/SUSE/saphanabootstrap-formula/blob/master/pillar.example

Theoretically, if a formula is used with the deployment pillar are set automatically.

But we are covering in the documentation the usage of a formula as standallone.

I do think that `zypper in *formula` and then provinding best practices is not enough.

I would feel totally left alone as user. I mean I could expect this maybe from a devel doc, but this isn't a devel doc.

I do think we need to think how we can fill the gap right now. E.g reading the doc, we should cover by scenario and say what needed technically to achieve it.

If we need to ease this from the devel side , I do think we need to do in order to improve the UX from enduser..

**Comment by tomschr** [comment]
@MalloZup Thanks Dario for your insight! Much appreciated. :+1: 

> I would feel totally left alone as user. I mean I could expect this maybe from a devel doc, but this isn't a devel doc.

Well, yes, the section and the procedure therein is a bit general. However, that was *on purpose* so it can fit a lot of use cases. That's why I wrote before the procedure: "The following procedure is written in a general way to make it fit for all use cases."

If I understand it right, you can't have specific steps unless you document a specific use case. The procedure is not perfect by any means. That's why I additionally wrote: "You need
some knowledge about Salt and how to customize it to your needs."

At the moment, I'm not sure how to further improve the doc.

> If we need to ease this from the devel side , I do think we need to do in order to improve the UX from end-user..

If you can make it simpler from the development side, I'd love to document that! :+1: However, for the moment, this seems more of a future topic. That goes beyond this PR.

If you have concrete suggestions on how to improve the doc, feel free to send me a review and I will include it. My plan was to merge it to master soon so we can publish it on Friday. We can always open a new PR and improve the doc. :wink: 

Would that be okay for you?

**Comment by MalloZup** [comment]
@tomschr I'm fine with this PR. I do have created 1 issue and jira task for following-up this.

**Review by MalloZup — APPROVED** [review]

**Comment by tomschr** [comment]
@MalloZup Thank you, much appreciated. :+1: 

@lproven Liam, could you do a language review? That would be fine! :+1:

File: xml/art_solution_automation.xml [hunk 7]
```diff
@@ -166,6 +166,163 @@ https://confluence.suse.com/x/8IEnGw
...
+
+    <para>
+      The following procedure is written in a general way to make it fit for all use cases.
+      You need some knowledge about &salt; and how to customize it to your needs.
+    </para>
+
+    <procedure xml:id="pro-sol-autom-hana-apply-salt-formulas">
+      <title>Applying &salt; Formulas</title>
+      <para>
+        In this procedure we show how to create a pillar structure for
```
Comment (lproven) [inline xml/art_solution_automation.xml:202]:
```suggestion
        In this procedure, we show how to create a pillar structure for
```

File: xml/art_solution_automation.xml [hunk 8]
```diff
@@ -166,6 +166,163 @@ https://confluence.suse.com/x/8IEnGw
...
+          Copy all pillar files to the right directory:
+        </para>
+        <screen>&prompt.sudo;<command>cp</command> -vi hana.sls cluster.sls /srv/pillar/hana
+&prompt.sudo;<command>cp</command> -vi netweaver.sls cluster.sls /srv/pillar/netweaver
+&prompt.sudo;<command>cp</command> -vi drbd.sls cluster.sls /srv/pillar/drbd</screen>
+      </step>
+      <step>
+        <para>
+          Modify the content of the pillar files depending on your needed configuration
+          of each &salt; formula.
```
Comment (lproven) [inline xml/art_solution_automation.xml:278]: This reads a little strangely to me.

How about:

         Modify the content of the pillar files for each &salt; formula, according to the configuration you need.

File: xml/art_solution_automation.xml [hunk 9]
```diff
@@ -166,6 +166,163 @@ https://confluence.suse.com/x/8IEnGw
...
+        </para>
+        <para>
+          Notice, in this example, we have three different
```
Comment (lproven) [inline xml/art_solution_automation.xml:281]:
```suggestion
          Notice that, in this example, we have three different
```

File: xml/art_solution_automation.xml [hunk 10]
```diff
@@ -166,6 +166,163 @@ https://confluence.suse.com/x/8IEnGw
...
+    'netweaver01,netweaver02':
+        - netweaver
+        - cluster
+
+    'drbd01,drbd02':
+        - drbd
+        - cluster</screen>
+        <para>
+          In the above example, folder part is removed, as we are pointing to the formulas with
+          their name.
```
Comment (lproven) [inline xml/art_solution_automation.xml:304]:
```suggestion
          their names.
```

File: xml/art_solution_automation.xml [hunk 11]
```diff
@@ -166,6 +166,163 @@ https://confluence.suse.com/x/8IEnGw
...
+        </para>
+      </step>
+      <step>
+        <para> Execute &salt; as shown in the following commands. Keep in mind,
+          &salt; supports more options, it is just an example: </para>
```
Comment (lproven) [inline xml/art_solution_automation.xml:309]:
```suggestion
        <para>
         Execute &salt; as shown in the following commands. Bear in mind,
         these are just examples&mdash;&salt; supports more options.
        </para>
```

File: xml/art_solution_automation.xml [hunk 12]
```diff
@@ -166,6 +166,163 @@ https://confluence.suse.com/x/8IEnGw
...
+        <stepalternatives>
+          <step>
+            <para>If you use a &saltmaster; with &salt; client, run:</para>
```
Comment (lproven) [inline xml/art_solution_automation.xml:312]:
```suggestion
            <para>If you use a &saltmaster; with &salt; clients, run:</para>
```

File: xml/art_solution_automation.xml [hunk 13]
```diff
@@ -166,6 +166,163 @@ https://confluence.suse.com/x/8IEnGw
...
+            <screen>&prompt.root;<command>salt</command> '*' state.highstate</screen>
+          </step>
+          <step>
+            <para>If you use a &salt; client only, run:</para>
```
Comment (lproven) [inline xml/art_solution_automation.xml:316]:
```suggestion
            <para>If you use only a &salt; client, run:</para>
```

File: xml/common_intro_feedback.xml [hunk 14]
```diff
@@ -15,7 +15,7 @@
  <title>Giving Feedback</title>
 
  <para>
-  We welcome feedback on and contributions to this documentation.
+  We welcome feedback for and contributions to this documentation.
```
Comment (lproven) [inline xml/common_intro_feedback.xml:18]: TBH, I preferred "on".

**Review by lproven — CHANGES_REQUESTED** [review]
In addition, in some pre-existing text:

line 26:
> The solutions that is described here works from &slsreg;&nbsp;12 SP3
That _are_ described here

File: xml/common_intro_feedback.xml (hunk 14 above)
Comment (tomschr) [reply xml/common_intro_feedback.xml]: Uups, that was by accident and should be committed. Fixed it with your suggestion in commit 5cbc528

File: xml/art_solution_automation.xml (hunk 8 above)
Comment (tomschr) [reply xml/art_solution_automation.xml]: Yep, fixed it with commit 5cbc528.

**Pull request merged**